command/git_push.go

```diff
@@ -0,0 +1,153 @@
+package command
+
+import (
+	"bytes"
+	"context"
+	"fmt"
+	"path/filepath"
+	"strings"
+
+	"github.com/evergreen-ci/evergreen/model"
+	"github.com/evergreen-ci/evergreen/rest/client"
+	restModel "github.com/evergreen-ci/evergreen/rest/model"
+	"github.com/mitchellh/mapstructure"
+	"github.com/mongodb/grip/level"
+	"github.com/pkg/errors"
+)
+
+// gitMerge is a command that commits tracked changes and pushes the resulting commit to a remote repository
+type gitPush struct {
+	// The root directory (locally) that the code is checked out into.
+	// Must be a valid non-blank directory name.
+	Directory string `plugin:"expand"`
+	base
+}
+
+func gitPushFactory() Command   { return &gitPush{} }
+func (c *gitPush) Name() string { return "git.push" }
+
+// ParseParams parses the command's configuration.
+// Fulfills the Command interface.
+func (c *gitPush) ParseParams(params map[string]interface{}) error {
+	err := mapstructure.Decode(params, c)
+	if err != nil {
+		return err
+	}
+
+	if c.Directory == "" {
+		return errors.Errorf("error parsing '%v' params: value for directory "+
+			"must not be blank", c.Name())
+	}
+
+	return nil
+}
+
+// Execute gets the source code required by the project
+func (c *gitPush) Execute(ctx context.Context, comm client.Communicator, logger client.LoggerProducer, conf *model.TaskConfig) error {
+	p, err := comm.GetTaskPatch(ctx, client.TaskData{ID: conf.Task.Id, Secret: conf.Task.Secret})
+	if err != nil {
+		return errors.Wrap(err, "Failed to get patch")
+	}
+
+	commands := getCheckoutBranchCommands(conf.ProjectRef.Branch)
+	if err = c.runCommands(ctx, commands, c.Directory, logger); err != nil {
+		return errors.Wrapf(err, "can't checkout '%s' branch", conf.ProjectRef.Branch)
+	}
+
+	// fail the merge if HEAD has moved
+	headSHA, err := c.revParse(ctx, logger, "HEAD")
+	if err != nil {
+		return errors.Wrap(err, "can't get SHA for HEAD")
+	}
+	if p.Githash != headSHA {
+		return errors.Errorf("tip of branch '%s' has moved. Expecting '%s', but found '%s'", conf.ProjectRef.Branch, p.Githash, headSHA)
+	}
+
+	// get author information
+	u, err := comm.GetUserAuthorInfo(ctx, p.Author)
+	if err != nil {
+		return errors.Wrapf(err, "can't get author information for user '%s'", p.Author)
+	}
+	displayName := restModel.FromAPIString(u.DisplayName)
+	email := restModel.FromAPIString(u.Email)
+
+	// push main patch
+	commands = getCommitCommands(displayName, email, p.Description, conf.ProjectRef.Branch)
+	if err = c.runCommands(ctx, commands, c.Directory, logger); err != nil {
+		return errors.WithStack(err)
+	}
+
+	// push module patches
+	for _, modulePatch := range p.Patches {
```

Comment:
Under what scenarios do you have more than the main patch here?

Reply:
if the patch includes changes to modules (the set-module subcommand of the commit queue command will upload patches to modules)